Build uniform grid in 1d fit

aewallwi opened the pull request:
Technically depends on #114  and #113. Reviewer should focuse on changes to `dspec.py` and `test_dspec.py`. This PR allows _fit_basis_1d to handle missing integrations by replacing them with flagged zeros on a uniform grid. Missing data causes the x-axis to not be uniformly spaced but it is spaced by integer multiples of a fundamental dx so we can fill in these gaps with flagged weight zero data to make the basis fitting behave and then throw away these points when we return the model fits.

r-pascua reviewed (changes requested):
This looks like a useful update that allows for easy processing of data that has gaps due to spurious dropouts; however, there are a few changes I would like to see to the new function (in particular, just some variable name changes and a bit of improvement of the documentation). In addition to the comments I left in this review, I think it would be nice for the new function to have a section of examples that show the user the expected behavior of the function.

On another note, I think that the testing changes should be their own PR. I have done a very large testing overhaul before, and it is a pretty monstrous process to get everything right, though it is very much worth the effort if the intention is to set a higher bar for testing. There is good progress on converting to ``pytest``-only, but there is still much work left to be done to improve the tests. I think that moving all the testing updates to a new branch and saving it for a different PR will serve to make the job of reviewing easier, keep the changes more modular, and let this PR get merged faster.

Thank you for all the work!

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+
+    pytest.raises(ValueError, dspec.gen_window, 'foo', 200)
+    # check Ncut ValueError
+    pytest.raises(ValueError, dspec.gen_window, 'bh', 200, edgecut_hi=101, edgecut_low=100)
```
This is now in a different PR.

aewallwi commented on uvtools/dspec.py:
```diff
@@ -109,6 +109,77 @@ def _get_filter_area(x, filter_center, filter_width):
+
+    Returns
+    -------
+        xout: array-like
+              If the separations on x are multiples of a single underlying minimum unit
+              returns x with all multiples of the fundamental unit filled in.
+              If x is already uniformly spaced, returns x unchanged. If separations are not
+              multiples of fundamental unit, also returns x unchanged.
+        yout: array-like
+              If the separations on x are multiples of a single underlying minimum unit
+              returns y with all multiples of the fundamental unit filled in with zeros.
+              If x is already uniformly spaced, returns y unchanged. If separations are not
+              multiples of fundamental unit, also returns y unchanged.
+        wout: array-like
+              If the separations on x are multiples of a single underlying minimum unit
+              returns w with all multiples of the fundamental unit filled in with zeros.
+              If x is already uniformly spaced, returns w unchanged. If separations are not
+              multiples of fundamental unit, also returns w unchanged.
+        inserted: array-like
+              boolean array indicating which x-values were inserted.
```
fixed.

aewallwi commented on uvtools/dspec.py:
```diff
@@ -109,6 +109,77 @@ def _get_filter_area(x, filter_center, filter_width):
+
+    Parameters
+    ----------
+    x: array-like,
+        array of x-values.
+    y: array-like,
+        array of y-values.
+    w: array-like,
+        array of weights.
+    xtol: float, optional.
+        fractional error tolerance to determine if x-values are
+        on an incomplete grid.
```
I've beefed up the docstring and added more descriptive variable (w -> weights, y -> data).

aewallwi commented on uvtools/dspec.py:
```diff
@@ -109,6 +109,77 @@ def _get_filter_area(x, filter_center, filter_width):
+              boolean array indicating which x-values were inserted.
+    """
+    xdiff = np.diff(x)
+    dx = np.abs(np.diff(x)).min() * np.sign(np.diff(x)[0])
```
yes.

aewallwi commented on uvtools/dspec.py:
```diff
@@ -109,6 +109,77 @@ def _get_filter_area(x, filter_center, filter_width):
+    for i in range(len(x) - 1):
+        grid_spacing =  (x[i + 1] - x[i]) / dx
+        integer_spacing = np.round(grid_spacing)
+        if not np.isclose(integer_spacing, grid_spacing, rtol=0, atol=xtol):
```
Thanks for catching this! I did indeed mean to have `atol = dx * xtol` and changed to this.

r-pascua replied:
I think that this change wasn't actually pushed yet? It looks like it still says `atol=xtol` on my end.

aewallwi replied:
Good catch. I've fixed this.

aewallwi commented on uvtools/dspec.py:
```diff
@@ -109,6 +109,77 @@ def _get_filter_area(x, filter_center, filter_width):
+            yout = y
+            wout = w
+            inserted = np.zeros(len(x), dtype=bool)
+            return xout, yout, wout, inserted
```
I've added coverage for this.

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1054,6 +1053,55 @@ def test_vis_clean():
+    xt = np.arange(0, 100) * 1.23157
+    yt = np.random.randn(len(xt)) + 1j * np.random.randn(len(xt))
+    wt = np.ones(len(xt))
+    for m in range(100):
```
make sure that we don't visit a special case. I've reduced this to two times.

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1054,6 +1053,55 @@ def test_vis_clean():
+        assert np.allclose(wout[to_remove], 0.0)
+        assert np.allclose(wout[to_keep], wt[to_keep])
+
+    # try negative grid.
+    xt = np.arange(0, 100) * -3.8271
+    yt = np.random.randn(len(xt)) + 1j * np.random.randn(len(xt))
+    wt = np.ones(len(xt))
```
Great tip! This will save me a lot of testing code in the future :).

aewallwi commented on uvtools/dspec.py:
```diff
@@ -109,6 +109,77 @@ def _get_filter_area(x, filter_center, filter_width):
+    nx_grid =int(np.round((x[-1] - x[0]) / dx)) + 1
+    xout = np.linspace(x[0], x[-1], nx_grid)
+    yout = np.zeros(xout.shape, dtype=np.complex128)
+    wout = np.zeros(xout.shape, dtype=np.float)
```
I've changed to this system.

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+    assert dspec.wedge_width(100., .001, 100, horizon=.1) == (2,-1)
+    assert dspec.wedge_width(100., .001, 100, horizon=.5) == (6,-5)
+    assert dspec.wedge_width(100., .001, 100, horizon=1.5) == (16,-15)
+    assert dspec.wedge_width(100., .001, 100, horizon=2.) == (21,-20)
```
Move this comment to #116

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+def test_delay_filter_2D():
+    NCHAN = 128
+    NTIMES = 10
+    TOL = 1e-6
```
Move this comment to #116

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+    wgts[0,:72] = 0.
+    dmdl, dres, info = dspec.delay_filter(data, wgts, 0., .1/NCHAN, tol=TOL, skip_wgt=0.5, mode='clean')
+    assert info['status']['axis_1'][0] == 'skipped'
+    assert info['status']['axis_1'][1] == 'success'
```
Move this comment to #116.

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+
+    # Test errors
+    pytest.raises(ValueError, dspec.fourier_model, 3, NFREQS)
+    pytest.raises(ValueError, dspec.fourier_model, np.empty((3, 3)), NFREQS)
```
Move this comment to #116

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+    pytest.raises(ValueError, dspec.delay_filter_leastsq_1d,
+                     data[0], flags[0], sigma, nmax=3, operator=np.empty((3, 3)))
+    pytest.raises(ValueError, dspec.delay_filter_leastsq_1d,
+                     data[0], flags[0], sigma, nmax=3, cn_guess=np.array([3]))
```
Move this comment #116

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+    # test multiple entries in filter_size
+    filter_size = (1e-2, 2e-2)
+    u, l = dspec.calc_width(filter_size, dt, nchan)
+    assert np.all((frs[u:l] < -1e-2) | (frs[u:l] > 2e-2))
```
Move this comment to #116

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+        win2 = dspec.gen_window(w, 100,normalization='mean')
+        assert np.all(np.isclose(win, win2*np.mean(win),atol=1e-6))
+        win3 = dspec.gen_window(w, 100,normalization='rms')
+        assert np.all(np.isclose(win, win3*np.sqrt(np.mean(win**2.)),atol=1e-6))
```
Move this comment to #116

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+    #check that skip_wgt is properly passed to clean
+    wgts[:72] = 0.
+    dmdl, dres, info = dspec.delay_filter(data, wgts, 0., .1/NCHAN, tol=TOL, skip_wgt=0.5, mode='clean')
+    assert info['status']['axis_1'][0] == 'skipped'
```
Move to #117

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1,211 +1,210 @@
+    np.testing.assert_allclose(dres[0,:], (data * wgts)[0,:], atol=NCHAN*TOL)
+    assert len(info['status']['axis_1']) == NTIMES
+    assert np.all([info['status']['axis_1'][i] == 'skipped' for i in list(info['status']['axis_1'])[:1]])
+    assert not np.any([info['status']['axis_1'][i] == 'skipped' for i in list(info['status']['axis_1'])[1:]])
```
Moved to #117

aewallwi commented:
Thanks for the review @r-pascua, especially the helpful pytest tips. I've restricted this PR to only address the 1d gridding stuff. I've moved the remainder of pytest migration to #116 (which is translating the existing tests verbatum into pytest compatible lines and tweaking the imports and test dir structure to accomodate codecov). I've moved the suggestions on reorganizing the existing tests (which is beyond the verbatum pytest translation) into their own issues which can be addressed in #117

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1054,6 +1053,55 @@ def test_vis_clean():
+        assert np.allclose(yout[to_keep], yt[to_keep])
+        assert np.allclose(yout[to_remove], 0.0)
+        assert np.allclose(wout[to_remove], 0.0)
+        assert np.allclose(wout[to_keep], wt[to_keep])
```
I've broken this test up.

aewallwi commented on uvtools/tests/test_dspec.py:
```diff
@@ -1092,7 +1140,21 @@ def test__fit_basis_1d():
-if __name__ == '__main__':
-    unittest.main()
+    assert np.allclose(mod5, mod6[to_keep])
+    assert np.allclose(resid5, resid6[to_keep])
```
I've broken this out.

r-pascua reviewed (approved):
I think the documentation is much clearer now, thank you! It looks like some changes you marked as resolved haven't been pushed yet--can you make sure that the remote branch is up-to-date with your local branch?

It looks like the new tests have improved relative to the original round of tests written, though the change to `.coveragerc` that was reverted seems to make it so that the coverage tool doesn't think that any of the new code is covered by tests. Is that being fixed in the PR dealing with migrating to `pytest`-only?

All that said, I think I'm happy to approve the changes. I trust that you'll get the minor changes I mentioned pushed, so go ahead and merge the PR once that's finished. Thank you for all your work!
